File: LeetCode/Python/Easy/20-Valid-Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack: deque[chr]=deque()

        for character in s:
            if character==')':
                opening: chr='('
                
                if self.foundMatching(stack, opening)==False:
                    return False
                else:
                    stack.pop()
            elif character==']':
                opening: chr='['

                if self.foundMatching(stack, opening)==False:
                    return False
                else:
                    stack.pop()
            elif character=='}':
                opening: chr='{'

                if self.foundMatching(stack, opening)==False:
                    return False
                else:
                    stack.pop()
            else:
                stack.append(character)
        
        return len(stack)==0

    def foundMatching(self, stack: deque[chr], opening: chr) -> bool:
        return stack and stack[-1]==opening if stack else False

File: LeetCode/Rust/Easy/20-Valid-Parentheses/solution.rs
impl Solution
{
    pub fn is_valid(s: String) -> bool
    {
        let mut stack: Vec<char>=Vec::new();

        for character in s.chars()
        {
            if character==')'
            {
                let matching: char='(';

                if Solution::contains_matching(&stack, &matching)==false
                {
                    return false;
                }
                else
                {
                    stack.pop();
                }
            }
            else if character==']'
            {
                let matching: char='[';

                if Solution::contains_matching(&stack, &matching)==false
                {
                    return false;
                }
                else
                {
                    stack.pop();
                }
            }
            else if character=='}'
            {
                let matching: char='{';

                if Solution::contains_matching(&stack, &matching)==false
                {
                    return false;
                }
                else
                {
                    stack.pop();
                }
            }
            else
            {
                stack.push(character);
            }
        }

        stack.is_empty()
    }

    pub fn contains_matching(stack: &Vec<char>, matching: &char) -> bool
    {
        if stack.is_empty()
        {
            return false;
        }

        let top: &char=stack.last().unwrap();

        top==matching
    }
}

impl Solution
{
    pub fn is_valid(s: String) -> bool
    {
        let mut stack: Vec<char>=Vec::new();

        for character in s.chars()
        {
            match(character)
            {
                '{' =>
                {
                    stack.push(character);
                },

                '[' =>
                {
                    stack.push(character);
                },

                '(' =>
                {
                    stack.push(character);
                },

                '}' =>
                {
                    if Some('{')!=stack.pop()
                    {
                        return false;
                    }
                },

                ']' =>
                {
                    if Some('[')!=stack.pop()
                    {
                        return false;
                    }
                },

                ')' =>
                {
                    if Some('(')!=stack.pop()
                    {
                        return false;
                    }
                },

                _ => (),
            }
        }

        stack.is_empty()
    }
}